stdlib/experimental/prometheus/from.go

```diff
@@ -0,0 +1,314 @@
+	spec, ok := prSpec.(*FromPrometheusProcedureSpec)
+	if !ok {
+		return nil, fmt.Errorf("invalid spec type %T", prSpec)
+	}
+
+	PrometheusIterator := PrometheusIterator{id: dsid, spec: spec, administration: a}
+
+	return execute.CreateSourceFromDecoder(&PrometheusIterator, dsid, a)
+}
+
+type PrometheusIterator struct {
+	id             execute.DatasetID
+	administration execute.Administration
+	spec           *FromPrometheusProcedureSpec
+	reader         *execute.RowReader
+
+	metrics []Metric
+	met     Metric
+	i       int
+	url     string
+	body    *http.Response
+}
+
+// Metric stores the fields that we need to construct Table
+type Metric struct {
+	Name      string                 `json:"name"`
+	Tags      map[string]string      `json:"tags"`
+	Fields    map[string]interface{} `json:"fields"`
+	Timestamp time.Time              `json:"timestamp"`
+}
+
+func (p *PrometheusIterator) Connect(ctx context.Context) error {
+
+	return nil
+}
+
+func (p *PrometheusIterator) parse(r io.Reader, header http.Header) (err error) {
+	var parser expfmt.TextParser
+	now := time.Now()
+
+	mediatype, params, err := mime.ParseMediaType(header.Get("Content-Type"))
+	if err != nil {
+		return err
+	}
+
+	// Prepare output
+	metricFamilies := make(map[string]*dto.MetricFamily)
+	if mediatype == "application/vnd.google.protobuf" &&
+		params["encoding"] == "delimited" &&
+		params["proto"] == "io.prometheus.client.MetricFamily" {
+		for {
+			mf := &dto.MetricFamily{}
+			if _, err := pbutil.ReadDelimited(r, mf); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return fmt.Errorf("reading metric family protocol buffer failed: %s", err)
+			}
+			metricFamilies[mf.GetName()] = mf
+		}
+	} else {
+		metricFamilies, err = parser.TextToMetricFamilies(r)
+		if err != nil {
+			return fmt.Errorf("reading text format failed: %s", err)
+		}
+	}
+
+	p.metrics = make([]Metric, 0)
+
+	// read metrics
+	for name, family := range metricFamilies {
+		for _, m := range family.Metric {
+			// read tags
+			tags := makeLabels(m)
+
+			// read fields
+			var fields map[string]interface{}
+
+			switch family.GetType() {
+			case dto.MetricType_SUMMARY:
+				continue // skip for now
+			case dto.MetricType_HISTOGRAM:
+				continue // skip for now
+			default: // counters, gagues and uptyped
+				fields = getNameAndValue(m)
+			}
+
+			// Store data needed to build flux Table
+			if len(fields) > 0 {
+				tm := now
+				if m.TimestampMs != nil && *m.TimestampMs > 0 {
+					tm = time.Unix(0, *m.TimestampMs*1000000)
+				}
+				met := Metric{
+					Timestamp: tm,
+					Tags:      tags,
+					Fields:    fields,
+					Name:      name,
+				}
+				p.metrics = append(p.metrics, met)
+			}
+		}
+	}
+	return nil
+}
+
+func (p *PrometheusIterator) Fetch(ctx context.Context) (bool, error) {
+	p.url = "http://demo.robustperception.io:9090/metrics"
+	resp, err := http.Get(p.url)
```

Comment:
This might be causing an infinite loop. `Fetch` is called repeatedly until it returns `false` and I'm not convinced this will ever return `false`.

It might be best to rewrite this to use the new `SourceIterator` which has an easier interface to implement.

Reply:
I haven't been experiencing an infinite loop here 🤔 There's a `return false, nil` at the bottom of this method when the conditional is satisfied. But you're right that I should update this to the new `SourceIterator`. I'll pull that in and see if helps with the speed.

Reply:
Ah! Now I understand what's going on. The reason this is slow is because you're making n http requests instead of making a single request and buffering the result while you read through it.

It would definitely be much easier to implement `SourceIterator` as it would make the code a lot more straightforward and less confusing.

Reply:
Okay, yeah that makes sense. I will try and implement `SourceIterator` here.

Reply:
Thanks Jonathan! I haven't implemented `SourceIterator` (I will try and do that before I open a real PR), but even separating the http.Get from the rest of the fetch logic made it much faster.